Add a section about CORS (#93)
connectrpc/connectrpc.com: docs/cors.md

response headers to JavaScript that are explicitly allowed.

Both Connect and gRPC-web require CORS preflights for cross-origin requests.
Note that we highly recommend to use a library like [rs/cors](https://github.com/rs/cors) 
for Go to implement the logic. This document gives you the necessary information
to configure CORS appropriately using any library.

## Preflight request

Here's an example for a CORS preflight request with the Connect protocol:

```
OPTIONS /connectrpc.greet.v1.GreetService/Greet HTTP/1.1
Origin: https://connectrpc.com
Access-Control-Request-Method: POST
Access-Control-Request-Headers: Connect-Protocol-Version,Content-Type
[other headers elided...]
```

Note that the browser transmits the origin, as well as the HTTP methods and 
header names of the actual request.

## Preflight response

Here is an example for a CORS preflight response for the preflight above that
allows all requests with the Connect and gRPC-web protocols, and asks the browser
to cache the information:

```
HTTP/1.1 204 No Content
Access-Control-Allow-Origin: https://connectrpc.com
Access-Control-Allow-Methods: POST, GET

> Access-Control-Allow-Headers: Content-Type,Connect-Protocol-Version,Connect-Timeout-Ms,Grpc-Timeout,X-Grpc-Web,X-User-Agent

Author: Was not deliberate. Removed the space here for consistency.

Access-Control-Max-Age: 7200
Vary: Origin,Access-Control-Request-Method,Access-Control-Request-Headers
[other headers elided...]
```

Let's take a look at the details:

### Allowing origins

`Access-Control-Allow-Origin: https://connectrpc.com` allows browsers to make 
requests from connectrpc.com via https with the default port 443. 

### Allowing methods and headers

> `Access-Control-Allow-Methods: POST, GET` is required to support gRPC-web and 

Author: Updated in https://github.com/connectrpc/connectrpc.com/pull/93/commits/a24a357c22daddc17a2a025d126123fdfc0dcbf3

Connect. gRPC-web will only use POST. Connect uses POST and optionally GET.

`Access-Control-Allow-Headers: Content-Type,Connect-Protocol-Version,Connect-Timeout-Ms,Grpc-Timeout,X-Grpc-Web,X-User-Agent`

> allows all necessary headers for Connect and gRPC-web.

Author: Updated in https://github.com/connectrpc/connectrpc.com/pull/93/commits/18711745ec65a41dbba37db8922a0922ae82491c

> **Note:** If your application uses custom request headers, you must explicitly 
> allow them. This includes commonly used headers such as `Authorization`.

### Caching preflight responses

> `Access-Control-Max-Age: 7200` lets browsers cache CORS information for longer, 

Author: Explained in https://github.com/connectrpc/connectrpc.com/pull/93/commits/fb805135d536c618ab34a2644c446774d3ab8b49

which reduces the number of preflight requests. Any changes to your CORS setup 
won't take effect until the cached data expires. Firefox caps this value at 24h, 
and modern Chrome caps it at 2h.

`Vary: Origin,Access-Control-Request-Method,Access-Control-Request-Headers` is 
used by the browser to cache CORS information based on the given header names.
To avoid poisoning intermediate caches, all request headers resulting in

> conditional preflight responses must be added to the `Vary` response header.

Author: Updated in https://github.com/connectrpc/connectrpc.com/pull/93/commits/33c2d3db885d7592445608c091a2d7bece769f22

{"greeting": "Hello, Buf!"}
```

All three exposed headers are necessary to support gRPC-web.

> **Note:** If your application uses custom response headers, you must explicitly
> expose them.  
> If your application uses custom response **trailers**, you must expose them with
> a `Trailer-` prefix, because Connect sends trailers for unary RPCs as headers.
> If your service sets the trailer `Bar`, the response must expose `Trailer-Bar`.

## Connect GET

For side-effect-free RPCs such as simple queries, Connect supports GET requests
that do not require CORS preflights. Browser will immediately issue the actual 
request, but there are two requirements:

1. No request headers must be set, including any application-specific header, or
  `Connect-Timeout-Ms` or `Grpc-Timeout`.
2. The response must set the header `Access-Control-Allow-Origin: *`.

If the first requirement is not met, a preflight request is issued. If the second
requirement is not met, the request will error in the browser.

## Avoid wildcards

While many CORS headers allow a wildcard `*`, they only do for requests without 
credentials, such as cookies, TLS client certificates, or authorization headers.

> We recommend to avoid wildcards, especially in a production environment. CORS 

Reviewer: ```suggestion
It's best to avoid wildcards in production. CORS 
```
